### PR TITLE
Reallocation UI

### PR DESCRIPTION
**Pull request by ryancreatescopy**

### CODE REVIEW

**Comment by samajammin** [comment]
@ryancreatescopy perhaps you can walk through this with @pettinarip sometime this week?

**Comment by ryancreatescopy** [comment]
Yep @pettinarip let me know when you have time to tackle this and we can go through what needs doing.

**Comment by pettinarip** [comment]
Ok, @ryancreatescopy. Tomorrow morning I'll ping you.

**Comment by ryancreatescopy** [comment]
Hi @pettinarip have tried to write out as many of the things I can think of that still need polishing in re: reallocation. 

Take a little look through these and then we can chat in a bit about anything that's not clear. 

Also, see if you can contribute/reallocate for yourself in your local – there may be stuff you notice that I haven't yet considered.

**TODOs**

- [x] Error: ``Your new total can't be more than your original ${formatAmount(this.contribution)} contribution.`` need to add the actual contribution amount as a legible number
- [x] Duplicate "time left" div – 1 should only display in read-only mode, the other in edit mode
- [x] Reallocate button shouldn't be visible (and it's time left) shouldn't be visible until you make an edit to the cart
- [x] Adding a new item to the cart should put the entire cart into edit mode (at the moment it is added in read-only mode)
- [x] New cart items should have a green background to demonstrate that they're new
- [x] If you add a new item to your cart and then "cancel" out of the edit mode it shouldn't be visible in your cart as the cart will now be in read-only state reflecting what's currently been transacted
- [x] ~~Split evenly function needs to be built (to be solved separately - see #120)~~
- [x] "Remove all" button shouldn't be visible unless cart is in edit mode. For this reason we should probably hide the `...` more button during reallocation
- [x] ~If a user reallocates, and some of their total goes to the matching pool, the read-only state should have an item for matching pool.~ For up-to-date designs for handling matching pool in cart, see: https://github.com/ethereum/clrfund/issues/106#issuecomment-856501552 Watch out: potential crossover with #104
- [x] #117 might not be in scope for this PR but still relevant that items you didn't offiially reallocate to should not be in your cart after quitting edit mode or if round transitions to tallying..
- [x] Decision needed: are we building the progress bars e.g:

image

**Comment by samajammin** [comment]
> New cart items should have a green background to demonstrate that they're new

The work @corwintines is exploring to add a `isCommitted` boolean on each `CartItem` may come in handy here...

**Comment by pettinarip** [comment]
Nice, ok. That makes things much easier xD. So, now when we load the items in the cart we show the list filtered by that boolean.

**Comment by pettinarip** [comment]
@samajammin @corwintines @wackerow Regarding this point

> Adding a new item to the cart should put the entire cart into edit mode (at the moment it is added in read-only mode)

I was thinking about how to calculate the `isEditMode` boolean that currently lives inside the Cart component.
I came up with that we might need to move this state into the store level so that we can change it whenever we add a new item to the cart (on the `ADD_CART_ITEM` mutation).
Why? because in the component level I can't know **when** exactly a new item was added.

Let me know if you think there is a different way we can handle this case.

**Comment by samajammin** [comment]
@pettinarip note #120 & #103 are related to this reallocation cart UI (but IMO can be addressed in a separate PRs).

**Comment by pettinarip** [comment]
Oh ok ok. Yes, I think it would be better to address them in separate PRs.

**Comment by samajammin** [comment]
>  Split evenly function needs to be built

I also think this should be done in a separate PR - I've updated #120 to include the logic of that functionality.

**Comment by ryancreatescopy** [comment]
Just a note to @pettinarip (cc. @samajammin)

We've changed our mind regarding how we show the matching pool total in the reallocation cart. Have updated the todo list above but for designs, check here: https://github.com/ethereum/clrfund/issues/106#issuecomment-856501552

Ultimately we'll be showing the matching pool only as a line item in the cart, not as a projectItem if that makes sense...


--- 

On a separate note, had a question that I just wanted to check:

After reallocating does the cart return to the read-only state (e.g. without button or reallocation message?) I believe it should. Here's what I mean...

image
e.g. in this screenshot I just reallocated, it’s confusing if the cart doesn’t return to a “saved” state. I believe the button and reallocation message shouldn't show at this point unless the user goes back into an edit state by pressing edit or adding an item to their cart…

Instead you should just see something like:

image
**(Don't worry about what the project items themselves look like in this PR, the design in this PR for read-only project Items is correct)**
And the way you get the action button to reappear is to elect to edit the cart by pressing 'edit' or adding a new item.

Does this make sense? I think this behaviour is maybe already present but just wanted to double check!

**Comment by pettinarip** [comment]
1. Matching pool design: ok, perfect. Not sure if Corwin is going to address that in the PR he is working. @corwintines please confirm. Otherwise, we could definitely work on that here. 
2. Yes, it totally makes sense. Let me double-check that. I thought that was working as you described.

**Comment by pettinarip** [comment]
@ryancreatescopy I just went over it again and here are some explanations:

1. Currently, in this PR, the Reallocate (submit) button and message:
	- display if the user is in edit mode
	- hide if the user cancels the edition
	- (I'm not sure how you reached that state that you showed in the screenshot but I couldn't reproduce that running this PR)

2. When the user Cancels the edit state, we should discard the editions and revert the original contributions made by the user (this is a requirement). This is something that is not possible to do right now. Corwin is working on that. Once he finishes with that, I'll be able to basically work on this and the rest of the TODOs that you mentioned in this comment: https://github.com/ethereum/clrfund/pull/107#issuecomment-854719722

**Comment by ryancreatescopy** [comment]
> * (I'm not sure how you reached that state that you showed in the screenshot but I couldn't reproduce that running this PR)
> *

To clarify, I didn't check this behaviour but just wanted to confirm what should be the behaviour.

**Comment by pettinarip** [comment]
Two key things I've added in this PR:
1. Moved the `editModeSelection` state from the Cart component to the store. Why? because I need to mutate/toggle it to a specific state when a different action is triggered. In this case, it's when the `ADD_CART_ITEM` action is executed. At that moment, we need to force it to be ON.
2. ~~Added an `isCommited` flag on each CartItem in the cart. Why? because I need to identify which items are committed and which are the newer ones in order to set a green background on the new ones. I didn't find a simpler/easier way of doing this.~~ Now, to know if the item is committed or new, it is being calculated in the CartItems component itself.

File: vue-app/src/store/index.ts [hunk 1]
```diff
@@ -187,8 +204,12 @@ export const mutations = {
     state.showCartPanel = !state.showCartPanel
   },
   [RESTORE_COMMITTED_CART_TO_LOCAL_CART](state) {
+    // Flag the committed items. Useful to identify new items added into the Cart
+    // during the reallocation phase.
+    const commitedCart = state.committedCart.map(item => ({ ...item, isCommited: true }))
```
Comment (pettinarip) [inline vue-app/src/store/index.ts]: ~~Here is the 2) point I was talking about. This flag is going to be used in the Cart template to set different styles in not committed items.~~

File: vue-app/src/store/index.ts [hunk 2]
```diff
@@ -147,6 +161,9 @@ export const mutations = {
     } else {
       throw new Error('item is already in the cart')
     }
+
+    // When a new item is added force edit mode ON
+    state.editModeSelection = true
```
Comment (pettinarip) [inline vue-app/src/store/index.ts]: Here is the 1) point where edit mode needs to be ON when a new item is added.

File: vue-app/src/store/index.ts (hunk 1 above)
Comment (pettinarip) [reply vue-app/src/store/index.ts]: Not used anymore. Now, to know if the item is committed or new, it is being calculated in the CartItems component itself.

File: vue-app/src/components/Cart.vue [hunk 3]
```diff
@@ -49,14 +49,13 @@
...
           <div class="flex-row-reallocation">
             <div class="semi-bold">{{editModeSelection ? "Edit contributions" : "Your contributions"}}</div>
-            <div class="semi-bold" @click="handleEditState" v-if="$store.getters.canUserReallocate">
-              <span v-if="editModeSelection">Cancel</span>
-              <span v-else>Edit</span>
+            <div class="semi-bold" v-if="$store.getters.canUserReallocate">
+              <button @click="handleEditState">{{ editModeSelection ? "Cancel" : "Edit" }}</button>
             </div>
-            <!-- <div @click="removeAll" v-if="$store.getters.canUserReallocate">Remove all</div> -->
+            <!-- <div v-if="editModeSelection" @click="removeAll">Remove all</div> -->
```
Comment (pettinarip) [inline vue-app/src/components/Cart.vue]: @samajammin the Remove All (not the one that lives in the dropdown) was commented by Ryan, so I leave it like that. As we discussed before, I think we can leave it in that state and tackle it down in a separate PR.

Comment (samajammin) [reply vue-app/src/components/Cart.vue]: Yep, let's remove that.

**Comment by samajammin** [comment]
@pettinarip just to confirm the question of the dropdown edit behavior - let's:
1. always display it - even when not in edit mode (like we do currently)
2. if an options is selected (i.e. "split evenly" or "remove all"), that should toggle editMode

File: vue-app/src/components/CartItems.vue [hunk 4]
```diff
@@ -25,6 +30,7 @@
           {{item.amount}} {{tokenSymbol}}
         </div>
       </div>
+      <!-- TODO: if in tallying phase and the user contributed to matching pool as part of reallocation we should list it here -->
```
Comment (samajammin) [inline vue-app/src/components/CartItems.vue]: @pettinarip is this a Ryan comment? Not really sure what this means.

Comment (pettinarip) [reply vue-app/src/components/CartItems.vue]: Yes, it's from Ryan. I don't know either. Maybe it is related to the "old" design where we were going to have a Matching pool item?

File: vue-app/src/components/Cart.vue [hunk 5]
```diff
@@ -49,14 +49,12 @@
           <div>Choose some projects that you want to contribute to...</div>
           <router-link to="/projects" class="btn-secondary mobile mt1">See projects</router-link>
         </div>
-        <div v-else-if="$store.getters.canUserReallocate && !isCartEmpty">
+        <div v-else-if="$store.getters.canUserReallocate && !isCartEmpty && !$store.getters.hasReallocationPhaseEnded">
```
Comment (samajammin) [inline vue-app/src/components/Cart.vue]: This addition is redundant - if `$store.getters.canUserReallocate` is `true`, then `!$store.getters.hasReallocationPhaseEnded` is also `true`.

**Review by samajammin — APPROVED** [review]
Nice work @pettinarip! Added a couple small commits.

**Pull request merged**